File: backend/src/main/java/cn/etarch/mao/harness/delegate/SubagentRecoveryResultFactory.java
package cn.etarch.mao.harness.delegate;

import cn.etarch.mao.harness.delegate.entity.SubagentExecution;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.LinkedHashMap;
import java.util.Map;

@Component
@RequiredArgsConstructor
public class SubagentRecoveryResultFactory {

    private final ObjectMapper objectMapper;

    public String create(SubagentExecution execution) {
        boolean completed = "COMPLETED".equals(execution.getStatus());
        boolean cancelled = "CANCELLED".equals(execution.getStatus());
        String result = execution.getResult();
        if (result == null || result.isBlank()) {
            result = completed ? "(子代理未产生文本输出)" : "子代理恢复失败：执行未产生结果";
        }

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("success", completed);
        response.put("cancelled", cancelled);
        response.put("agent_type", execution.getAgentType());
        response.put("child_session_id", execution.getChildSessionId());
        response.put("result", result);
        if (!completed) {
            response.put("error", result);
        }
        response.put("rounds", valueOrZero(execution.getTotalRounds()));
        response.put("tool_calls", valueOrZero(execution.getTotalToolCalls()));
        response.put("usage", Map.of(
                "prompt_tokens", valueOrZero(execution.getTotalPromptTokens()),
                "completion_tokens", valueOrZero(execution.getTotalCompletionTokens()),
                "total_tokens", valueOrZero(execution.getTotalPromptTokens())
                        + valueOrZero(execution.getTotalCompletionTokens())));
        if ("FOLLOWUP".equals(execution.getInvocationType())) {
            response.put("follow_up", true);
        }
        try {
            return objectMapper.writeValueAsString(response);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException("Failed to serialize recovered subagent result", e);
        }
    }

    private int valueOrZero(Integer value) {
        return value != null ? value : 0;
    }
}

File: backend-ts/src/harness/delegate/subagent-recovery-result-factory.ts
import type { SubagentExecution } from '../../session/types.js';

export class SubagentRecoveryResultFactory {
  static build(execution: SubagentExecution): Record<string, unknown> {
    const cancelled = execution.status === 'CANCELLED';
    const success = execution.status === 'COMPLETED';
    const result = execution.result || (success ? '(子代理未产生文本输出)' : '子代理恢复失败');
    const payload: Record<string, unknown> = {
      success,
      cancelled,
      agent_type: execution.agentType ?? '',
      child_session_id: execution.childSessionId,
      result,
      rounds: execution.totalRounds ?? 0,
      tool_calls: execution.totalToolCalls ?? 0,
    };
    if (execution.invocationType === 'FOLLOWUP') payload.follow_up = true;
    if (!success) payload.error = result;
    const promptTokens = execution.totalPromptTokens ?? 0;
    const completionTokens = execution.totalCompletionTokens ?? 0;
    if (promptTokens > 0 || completionTokens > 0) {
      payload.usage = {
        prompt_tokens: promptTokens,
        completion_tokens: completionTokens,
        total_tokens: promptTokens + completionTokens,
      };
    }
    return payload;
  }

  static invocationArguments(execution: SubagentExecution): Record<string, unknown> {
    return execution.invocationType === 'FOLLOWUP'
      ? { child_session_id: execution.childSessionId, task: execution.taskDescription ?? '' }
      : { agent_type: execution.agentType ?? '', task: execution.taskDescription ?? '' };
  }
}
